### PR TITLE
WSTEAM1-476 - Show Russian email promotion banner on Optimo articles (V1)

### PR DESCRIPTION
**Pull request by Isabella-Mitchell**
Resolves JIRA [WSTEAM1-476](https://jira.dev.bbc.co.uk/browse/WSTEAM1-476)

Overall changes
======
- Show Russian email promotion banner on Optimo articles, in addition to the CPSAssets where it is already being shown.
- Update `disclaimer` component following component standards and new UX/ A11y requirments. 

- Note V1 (or dirty version) refers to the fact we are using toggles in Simorgh. The ultimate goal is to use toggles in the BFF so that we can insert the disclaimer into the BFF response.

Example on Optimo page:

Screenshot 2023-07-04 at 16 18 32

Code changes
======
**data/...**
- Update fixture data for articles & cpsAssets so that they mirror the BFF response and can be rendered locally for testing.
- Add a new article for the Russian service.

**src/app/components**
- Update Disclaimer component based on component standards.
- Update Unit tests. (See Questions/ To Do)
- Update storybook file, create README and add metadata.

**src/app/pages/...**
- Update imports into necessary CPSAsset pages to use new disclaimer block
- Update Article Page to render disclaimer with new disclaimer block.

**src/app/routes/...**
- Update augmentWithDiclaimer function so that it works with both CPS and Optimo assets.
- Add position argument to augmentWithDiclaimer to decide whether it renders before the timestamp (Optimo) or after the timestamp (CPS)
- Update `article/getInitialData/index.ts` so that it renders the disclaimer on Optimo pages based on Toggles (this is a working version until we can use toggles in the BFF).

Questions/ Notes for Code Review
======
- Disclaimer unit tests - I feel like these should be updated to be more inline with new standards.
- Disclaimer story - These uses service config data as I'm unsure how to mock this.
- augmentWithTImestamp unit tests - I feel these should these be updated to reflect Optimo pages
- routes/article/getInitialData unit tests - should tests be added here to check the disclaimer is added?

Testing
======
1. Check out this PR `WSTEAM1-476`
2. Launch the development server with the command `yarn dev`
3. Visit the following links to check the behaviour matches the ACs on the ticket

(Should Render)
- Russian Article Page: http://localhost:7080/russian/articles/cjejv43ylr0o
- Russian Article Page (another example): http://localhost:7080/russian/articles/ck7pz7re3zgo
- Russian Article .amp Page: http://localhost:7080/russian/articles/cjejv43ylr0o.amp
- Russian Article .app Page: http://localhost:7080/russian/articles/cjejv43ylr0o.app
- Russian STY Page: http://localhost:7080/russian/news-55041160
- Russian STY .amp Page: http://localhost:7080/russian/features-54391793.amp
- Russian PGL Page: http://localhost:7080/russian/features-45782775
- Russian MAP Page: http://localhost:7080/russian/media-49281069

(Should not render)
- Non-Russian Article Page: http://localhost:7080/afrique/articles/cz216x22106o
- Non-Russian STY Page: http://localhost:7080/kyrgyz/23292889
- Non-Russian PGL Page: http://localhost:7080/azeri/azerbaijan-44208474
- Non-Russian MAP Page: http://localhost:7080/pidgin/23248703

5. View the component on Storybook - https://wsteam1-476--5d28eb3fe163f6002046d6fa.chromatic.com/?path=/story/new-components-disclaimer--disclaimer

An A11y swarm will be required for this ticket.

Helpful Links
======
_Add Links to useful resources related to this PR if applicable._
- Designs inc Screen Reader UX: https://www.figma.com/file/U5b55Sl34Vh6vDxRKMwev1/Russian-service-disclaimer-Optimo-PW-FINAL-20230622?type=design&node-id=113%3A379&mode=design&t=aliHubWDAfsBKFpa-1
- A11y ACs: https://paper.dropbox.com/doc/Information-panel-formerly-Russian-Disclaimer-accessibility-acceptance-criteria--B7aaG0aTtvJgbudZi_JkAbFEAg-kR9WZaIuJi73YyUIl3XKw

[Coding Standards](https://github.com/bbc/simorgh/blob/latest/docs/Coding-Standards/README.md)

[Repository use guidelines](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/repository-guidelines.md)

### CODE REVIEW

File: src/app/routes/article/getInitialData/index.ts [hunk 1]
```diff
@@ -19,6 +20,16 @@ type Props = {
   path: string;
   pageType: 'article' | 'cpsAsset';
   variant?: Variants;
+  toggles?: Toggles;
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const addDisclaimer = async (pageData: any, toggles: Toggles | undefined) => {
```
Comment (Isabella-Mitchell) [inline src/app/routes/article/getInitialData/index.ts]: Used a comment to stop this giving me a warning. Not sure if there is a type I can use here?

File: src/app/components/Disclaimer/index.stories.tsx [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+import { ToggleContextProvider } from '../../contexts/ToggleContext';
+import ThemeProvider from '../ThemeProvider';
+import { StoryProps } from '../../models/types/storybook';
+
+import metadata from './metadata.json';
+import DisclaimerComponent from '.';
+
+const Component = ({ service, variant }: StoryProps) => {
+  return (
+    <ToggleContextProvider
```
Comment (Isabella-Mitchell) [inline src/app/components/Disclaimer/index.stories.tsx:14]: wasn't sure how to mock the serviceConfig so this is taking disclaimer text and link etc from the russian serviceConfig.

File: src/app/components/Disclaimer/index.test.tsx [hunk 3]
```diff
@@ -43,60 +39,51 @@ const renderComponent = (
...
     expect(getAllByRole('link').length).toBe(3);
   });
-
   it('should not render the disclaimer when the disclaimer toggle is not enabled', () => {
     const { container } = renderComponent({ enabled: false });
     expect(container).toBeEmptyDOMElement();
   });
-
   it('should not render when disclaimer is null', () => {
+    // @ts-expect-error - not expecting null
```
Comment (Isabella-Mitchell) [inline src/app/components/Disclaimer/index.test.tsx:76]: Used comments in this file to stop errors. I think this could be avoided with a better use of types.

File: src/app/components/Disclaimer/index.test.tsx [hunk 4]
```diff
@@ -43,60 +39,51 @@ const renderComponent = (
         },
       }}
     >
+      {/* @ts-expect-error - to diagnose */}
```
Comment (Isabella-Mitchell) [inline src/app/components/Disclaimer/index.test.tsx:42]: I think this error is happening because the ServiceContext.Provider isn't expecting the value prop. The tests run and pass. I couldn't work out how to refactor this to use Types and also get the tests to pass.

File: src/app/components/Disclaimer/index.tsx [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+        >
+          {disclaimer &&
+            Object.values(disclaimer).map(para => {
+              const linkText: string | undefined = path(['text'], para);
+              const linkUrl: string | undefined = path(['url'], para);
+              return linkUrl ? (
+                <InlineLink
+                  className="focusIndicatorReducedWidth"
+                  css={styles.inlineLink}
+                  key={linkText as string}
```
Comment (Isabella-Mitchell) [inline src/app/components/Disclaimer/index.tsx]: Added these '... as string' to fix Typescript errors. Not sure this is the best approach.

File: src/app/components/Disclaimer/index.test.tsx (hunk 3 above)
Comment (karinathomasbbc) [reply src/app/components/Disclaimer/index.test.tsx:76]: I think this is OK. If you really wanted to get rid of expected errors, you could remove the test, since typescript is technically preventing you from passing a null value to your component.

File: src/app/components/Disclaimer/index.test.tsx [hunk 6]
```diff
@@ -43,60 +39,51 @@ const renderComponent = (
...
   });
-
   it('should not render when disclaimer is null', () => {
+    // @ts-expect-error - not expecting null
     const { container } = renderComponent({ enabled: true }, null);
     expect(container).toBeEmptyDOMElement();
   });
-
   it('should not render when disclaimer is empty object', () => {
+    // @ts-expect-error - not expecting empty object
```
Comment (karinathomasbbc) [inline src/app/components/Disclaimer/index.test.tsx:81]: As above - this is OK to keep, but you could remove the test if you wanted to.

As long as we give a reason why we're using ts-expect-error, I am happy :)

File: src/app/components/Disclaimer/index.tsx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+  const shouldShow = enabled && disclaimer && !isEmpty(disclaimer);
+
+  if (!shouldShow) return null;
+
+  const infoBannerLabelTranslation = pathOr(
+    'Information',
+    ['infoBannerLabel'],
+    translations,
+  );
```
Comment (karinathomasbbc) [inline src/app/components/Disclaimer/index.tsx]: Is there any way we could avoid using `pathOr`?

```suggestion
  const infoBannerLabelTranslation = translations?.infoBannerLabel || 'Information';
```

File: src/app/components/Disclaimer/index.tsx [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            styles.inner,
+            pageType === ARTICLE_PAGE && styles.increaseTopMargin,
+          ]}
+          size="longPrimer"
+          fontVariant="sansLight"
+        >
+          {disclaimer &&
+            Object.values(disclaimer).map(para => {
+              const linkText: string | undefined = path(['text'], para);
+              const linkUrl: string | undefined = path(['url'], para);
```
Comment (karinathomasbbc) [inline src/app/components/Disclaimer/index.tsx]: As above, can we avoid using `path` - which should then remove the need for typing here and casting to a string below

File: src/app/pages/PhotoGalleryPage/PhotoGalleryPage.jsx [hunk 9]
```diff
@@ -38,6 +37,7 @@ import ATIAnalytics from '../../components/ATIAnalytics';
 import ChartbeatAnalytics from '../../components/ChartbeatAnalytics';
 import LinkedData from '../../components/LinkedData';
 import { ServiceContext } from '../../contexts/ServiceContext';
+import disclaimer from '../../components/Disclaimer';
```
Comment (karinathomasbbc) [inline src/app/pages/PhotoGalleryPage/PhotoGalleryPage.jsx]: **nitpick** (as I realise it was already like this): I thought component were supposed to be in capital case?

File: src/app/routes/cpsAsset/getInitialData/augmentWithDisclaimer/index.js [hunk 10]
```diff
@@ -11,18 +11,23 @@ const disclaimerBlock = {
...
   );
 
-export default toggles => pageData =>
-  isDisclaimerToggledOn(toggles) ? insertDisclaimer(pageData) : pageData;
+const augmentWithDisclaimer = (toggles, positionFromTimestamp) => pageData =>
+  isDisclaimerToggledOn(toggles)
+    ? insertDisclaimer(pageData, positionFromTimestamp)
+    : pageData;
+
+export default augmentWithDisclaimer;
```
Comment (karinathomasbbc) [inline src/app/routes/cpsAsset/getInitialData/augmentWithDisclaimer/index.js]: This could have been left as the default export...
```suggestion
export default (toggles, positionFromTimestamp) => pageData =>
  isDisclaimerToggledOn(toggles)
    ? insertDisclaimer(pageData, positionFromTimestamp)
    : pageData;
```

File: src/app/routes/article/getInitialData/index.ts (hunk 1 above)
Comment (karinathomasbbc) [reply src/app/routes/article/getInitialData/index.ts]: There isn't... yet. I do wonder if at some point we need to bite the bullet and add a global pageData definition. (not in this PR)

File: src/app/components/Disclaimer/index.test.tsx [hunk 11]
```diff
@@ -43,60 +39,51 @@ const renderComponent = (
...
     const { container } = renderComponent({ enabled: true }, {});
     expect(container).toBeEmptyDOMElement();
   });
-
   it('should render links with external label if links are external', () => {
     const { getByLabelText } = renderComponent();
     expect(getByLabelText('IOS, внешняя')).toBeInTheDocument();
   });
-
-  it('should not render links with external label if isExternal is false', () => {
```
Comment (karinathomasbbc) [inline src/app/components/Disclaimer/index.test.tsx:98]: Did we mean to delete this test?

File: src/app/components/Disclaimer/index.test.tsx (hunk 4 above)
Comment (karinathomasbbc) [reply src/app/components/Disclaimer/index.test.tsx:42]: Are you able to use ServiceContextProvider with props instead?

File: src/app/components/Disclaimer/index.stories.tsx [hunk 12]
```diff
@@ -0,0 +1,40 @@
...
+    </ToggleContextProvider>
+  );
+};
+
+export default {
+  title: 'Components/Disclaimer',
+  Component,
+  parameters: {
+    metadata,
+    chromatic: { disable: true },
```
Comment (karinathomasbbc) [inline src/app/components/Disclaimer/index.stories.tsx]: Do we mean to omit the disclaimer component from the Chromatic (visual) snapshots?

File: src/app/components/Disclaimer/index.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        </ThemeProvider>
+      </ServiceContextProvider>
+    </ToggleContextProvider>
+  );
+};
+
+export default {
+  title: 'Components/Disclaimer',
+  Component,
+  parameters: {
```
Comment (karinathomasbbc) [inline src/app/components/Disclaimer/index.stories.tsx:33]: Is there a readme for this component? Should we consider adding it as part of this work? (and also include it in Storybook)

File: src/app/components/Disclaimer/index.styles.tsx
Comment (karinathomasbbc) [inline src/app/components/Disclaimer/index.styles.tsx:1]: Technically this file does not need to be tsx as there is no React - can it be renamed to `index.styles.ts`

File: src/app/components/Disclaimer/index.tsx [hunk 14]
```diff
@@ -0,0 +1,78 @@
+/** @jsxRuntime classic */
+/** @jsx jsx */
+import { jsx } from '@emotion/react';
+import { useContext, PropsWithChildren } from 'react';
+import useToggle from '#hooks/useToggle';
+import { ServiceContext } from '#contexts/ServiceContext';
+import { RequestContext } from '#contexts/RequestContext';
+import { GridItemLarge } from '#components/Grid';
```
Comment (karinathomasbbc) [inline src/app/components/Disclaimer/index.tsx]: While we're in here, could we update the imports to use a relative path instead of the # alias, as per the React Component Standards?

**Review by karinathomasbbc — COMMENTED** [review]
Great work on migrating this component - I know how tricky it can be 🙃 

I've tried to respond to some of your comments, and have a couple of suggestions / questions on a few other things.

File: src/app/components/Disclaimer/index.tsx [hunk 15]
```diff
@@ -0,0 +1,78 @@
...
+    <GridItemLarge>
+      <section
+        css={[
+          styles.infoBanner,
+          increasePaddingOnDesktop && styles.increasePaddingOnDesktop,
+        ]}
+        role="region"
+        aria-label={infoBannerLabelTranslation}
+      >
+        <Paragraph
```
Comment (greenc05) [inline src/app/components/Disclaimer/index.tsx]: Could we please change this to a strong element as per the screen reader UX?

File: src/app/components/Disclaimer/index.stories.tsx (hunk 13 above)
Comment (greenc05) [reply src/app/components/Disclaimer/index.stories.tsx:32]: Along with the health check component?

File: src/app/components/Disclaimer/index.styles.tsx [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+    css({
+      [mq.GROUP_4_MIN_WIDTH]: {
+        padding: `${spacings.DOUBLE}rem ${spacings.QUINTUPLE}rem`,
+      },
+    }),
+
+  inlineLink: ({ palette }: Theme) =>
+    css({
+      color: palette.GREY_10,
+      borderBottom: `${pixelsToRem(1)}rem solid ${palette.GREY_10}`,
```
Comment (greenc05) [inline src/app/components/Disclaimer/index.styles.tsx]: Should we have some shared styles we can pull in for an inline link?

**Review by greenc05 — COMMENTED** [review]
Looking good, just a minor comment 😄

**Comment by greenc05** [comment]
Also noticed you don't have the final figma file (out of date one above), could you check it's as per the final UX spec please? I believe the text shouuld be regular rather than light weight....

https://www.figma.com/file/U5b55Sl34Vh6vDxRKMwev1/Information-panel-(Russian-Disclaimer)-Optimo-PW-FINAL-20230622?type=design&node-id=203-450&mode=design&t=Kc3FIBkJFaIVU78M-0

File: src/app/routes/cpsAsset/getInitialData/augmentWithDisclaimer/index.js (hunk 10 above)
Comment (Isabella-Mitchell) [reply src/app/routes/cpsAsset/getInitialData/augmentWithDisclaimer/index.js]: Ah cool thanks! I think got in a bit of an export muddle 🫠

File: src/app/components/Disclaimer/index.test.tsx (hunk 11 above)
Comment (Isabella-Mitchell) [reply src/app/components/Disclaimer/index.test.tsx:98]: I'm using the newer inlineLink component which doesn't seem to take the isExternal prop. (If I understand correctly, it's worked out here: https://github.com/bbc/simorgh/blob/latest/src/app/components/InlineLink/index.tsx#L39

So I didn't think the test was relevant.

File: src/app/components/Disclaimer/index.tsx (hunk 15 above)
Comment (Isabella-Mitchell) [reply src/app/components/Disclaimer/index.tsx]: Sorry I missed this! Will update

File: src/app/components/Disclaimer/index.styles.tsx (hunk 16 above)
Comment (Isabella-Mitchell) [reply src/app/components/Disclaimer/index.styles.tsx]: So the inline link has default styles but the colours are different for this component. E.g. the default colour for the inline link is palette.EBON, but for this component it is palette.GREY_10.

There is some duplication where I have declared the focus/ hover styles in postbox, as this is the same on the default inline link. However, if I remove this declaration, then it appears the 'visited' colour declaration overrides the inline link's focus/ hover states. So it seemed logical to duplicate the styles here.

File: src/app/components/Disclaimer/index.test.tsx (hunk 4 above)
Comment (Isabella-Mitchell) [reply src/app/components/Disclaimer/index.test.tsx:42]: No I couldn't get that to work either. If I change  ServiceContext.Provider to  ServiceContextProvider then all the tests fail. I think this is because ServiceContextProvider is only expecting props: `children`, `service`, `variant`, `pageLang`. 

I don't know how to refactor this to pass in the Disclaimer Fixture data, since that is taken from the Config files and not passed in to the component as content.

File: src/app/components/Disclaimer/index.stories.tsx (hunk 12 above)
Comment (Isabella-Mitchell) [reply src/app/components/Disclaimer/index.stories.tsx]: Ah no that was just copied across from the legacy component

**Review by shayneahchoon — APPROVED** [review]

File: src/app/routes/cpsAsset/getInitialData/augmentWithDisclaimer/index.js [hunk 17]
```diff
@@ -11,18 +11,21 @@ const disclaimerBlock = {
...
     ),
     pageData,
   );
 
-export default toggles => pageData =>
-  isDisclaimerToggledOn(toggles) ? insertDisclaimer(pageData) : pageData;
+export default (toggles, positionFromTimestamp) => pageData =>
+  isDisclaimerToggledOn(toggles)
+    ? insertDisclaimer(pageData, positionFromTimestamp)
+    : pageData;
```
Comment (HarveyPeachey) [inline src/app/routes/cpsAsset/getInitialData/augmentWithDisclaimer/index.js]: Do we need to do some form of sfv check here for the requirements? Or is that going to be done in another PR? We have a utility for it if that's the case [here](https://github.com/bbc/simorgh/blob/1a8f8eb6bbb1a6e8b8e4342f177a323aee0a94ed/src/app/routes/article/utils/paramChecks.ts#L14). https://jira.dev.bbc.co.uk/browse/WSTEAM1-464#:~:text=Article%20SFV%20(Video,flag%20is%20set)

File: src/app/routes/cpsAsset/getInitialData/index.js [hunk 18]
```diff
@@ -54,7 +54,7 @@ const processOptimoBlocks = toggles =>
     augmentWithTimestamp,
     only(
       [MEDIA_ASSET_PAGE, STORY_PAGE, PHOTO_GALLERY_PAGE],
```
Comment (HarveyPeachey) [inline src/app/routes/cpsAsset/getInitialData/index.js:56]: Do we need to add [ARTICLE_PAGE](https://github.com/bbc/simorgh/blob/1a8f8eb6bbb1a6e8b8e4342f177a323aee0a94ed/src/app/routes/utils/pageTypes.ts#L1) here? I guess this might relate to my sfv comment, does this mean it won't render for MEDIA_ARTICLE_PAGE because it's not present in this check?

File: src/app/routes/article/getInitialData/index.ts [hunk 19]
```diff
@@ -19,6 +20,16 @@ type Props = {
...
+  try {
+    return augmentWithDisclaimer(toggles, 0)(pageData);
```
Comment (HarveyPeachey) [inline src/app/routes/article/getInitialData/index.ts]: Could we make the second argument here a named const or define an [object parameter](https://github.com/bbc/simorgh/blob/1a8f8eb6bbb1a6e8b8e4342f177a323aee0a94ed/docs/Coding-Standards/Clean-Code.stories.mdx#L185-L220) in the function definition for `augmentWithDisclaimer` so it's more clear what this number does?

**Review by HarveyPeachey — COMMENTED** [review]
Looks really good, just a couple of comments

File: src/app/routes/cpsAsset/getInitialData/augmentWithDisclaimer/index.js (hunk 17 above)
Comment (Isabella-Mitchell) [reply src/app/routes/cpsAsset/getInitialData/augmentWithDisclaimer/index.js]: I assume article SFV is referring to Media Article Pages. So this would be handled in `src/app/routes/article/getInitialData/index.ts`

But I see your point. Based in the logic in that file, it seems likely that the disclaimer is being added into the Media Article Page response. (I don't have any test assets to check this)

I don't think the disclaimer would actually be rendered on the Media Article Page due to the Disclaimer component not being included in the 'componentsToRender' function. I have added this to [Article Pages](https://github.com/bbc/simorgh/blob/b03ca32e78fb053d2634c4c6e394be1df656b44c/src/app/pages/MediaAssetPage/MediaAssetPage.jsx#L164) to make it render.

But obviously we don't want to be adding stuff we're not not rendering, so I'll add the logic to stop that happening.

File: src/app/routes/cpsAsset/getInitialData/index.js (hunk 18 above)
Comment (Isabella-Mitchell) [reply src/app/routes/cpsAsset/getInitialData/index.js:56]: See my previous comment - I think this type of check should be present in `src/app/routes/article/getInitialData/index.ts`

File: src/app/routes/article/getInitialData/index.ts (hunk 19 above)
Comment (Isabella-Mitchell) [reply src/app/routes/article/getInitialData/index.ts]: Good thinking!

File: src/app/routes/cpsAsset/getInitialData/augmentWithDisclaimer/index.js (hunk 17 above)
Comment (Isabella-Mitchell) [reply src/app/routes/cpsAsset/getInitialData/augmentWithDisclaimer/index.js]: I've updated this and added some unit tests in to: `src/app/routes/article/utils/addDisclaimer.test.ts`

**Review by HarveyPeachey — APPROVED** [review]
